File: PMD/EventBuffer.py

import numpy as np

class EventBuffer:
    def __init__(self, width, height, depth, types):
        # create empty dict if no types passed
        if types is None:
            types = {}
        # Set data types
        self._timestamp_t = types.get('timestamp_t', 'u8')
        self._cluster_id_t = types.get('cluster_id_t', 'u2')
        self._cluster_weight_t = types.get('cluster_weight_t', 'u4')
        self._buffer_depth_t = types.get('buffer_depth_t', 'u1')

        # Initialize
        self._unassigned = np.iinfo(np.dtype(self._cluster_id_t)).max
        self._max_clusters = self._unassigned + 1
        self._ts_buffer = np.zeros((width, height, depth), dtype=self._timestamp_t)
        self._id_buffer = np.full((width, height, depth), self._unassigned, dtype=self._cluster_id_t)
        self._top = np.zeros((width, height), dtype=self._buffer_depth_t)
        self._width = width
        self._height = height
        self._depth = depth

    # look in the vicinity of (x, y), count nearby events and get all nearby clusters
    def check_vicinity(self, x, y, t, tf, tc):
        # clip indices to the edges of the buffer
        x_range = slice(max(x-1, 0), min(x+2, self._width))
        y_range = slice(max(y-1, 0), min(y+2, self._height))

        ts_buffer_slice = self._ts_buffer[x_range, y_range, :]
        id_buffer_slice = self._id_buffer[x_range, y_range, :]
        # count the number of correlated events within tf
        num_correlated = np.count_nonzero(ts_buffer_slice > t-tf)
        # get a sorted list of nearby cluster ids
        clusters = np.unique(id_buffer_slice[ts_buffer_slice > t-tc])[:-1]

        return num_correlated, clusters
    
    def get_flat_id_buffer(self):
        top = np.transpose(np.squeeze(np.take_along_axis(self._id_buffer, self._top[:, :, np.newaxis], axis=2)))
        assigned = np.nonzero(top!=self._unassigned)
        return top, assigned

    def get_ts_buffer_top(self, x=None, y=None):
        return self._id_buffer[x, y, self._top[x, y]]

    def add_event(self, x, y, t, cluster=None):
        self._top[x, y] = (self._top[x, y] + 1) % self._depth

        if self._id_buffer[x, y, self._top[x, y]] == self._unassigned:
            u_ids = np.array([])
            u_x = np.array([])
            u_y = np.array([])
        else:
            u_ids = np.array([self._id_buffer[x, y, self._top[x, y]]])
            u_x = np.array([x])
            u_y = np.array([y])

        self._ts_buffer[x, y, self._top[x, y]] = t
        self._id_buffer[x, y, self._top[x, y]] = self._unassigned if cluster is None else cluster

        return u_ids, u_x, u_y

    def remove_expired_events(self, domain, t):
        x_range, y_range = domain
        
        ts_buffer_slice = self._ts_buffer[x_range, y_range, :]
        id_buffer_slice = self._id_buffer[x_range, y_range, :]

        expired = ts_buffer_slice < t
        assigned = id_buffer_slice != self._unassigned
        to_remove = np.where(np.logical_and(expired, assigned))

        u_ids = np.array(id_buffer_slice[to_remove], copy=True)
        
        id_buffer_slice[to_remove] = self._unassigned

        self._id_buffer[x_range, y_range, :] = id_buffer_slice

        u_x = np.add(x_range.start, to_remove[0])
        u_y = np.add(y_range.start, to_remove[1])

        return u_ids, u_x, u_y

    def get_cluster_map(self, id):
        return np.any(self._id_buffer == id, 2)

    def merge_clusters(self, ids):
        target = ids[0]
        others = ids[1:]

        # print("assigned ", others, "to ", target)

        for id in others:
            # print("id", id, "buffer", self._id_buffer[self._id_buffer == id])
            self._id_buffer[self._id_buffer == id] = target

File: PMD/EventBuffer.cpp

#include "EventBuffer.h"
#include "ClusterBuffer.h"

namespace PMD {

    size_t EventBuffer::buffer_queue::depth = 0;

    EventBuffer::EventBuffer(size_t width, size_t height, size_t depth, ClusterBuffer &cluster_buffer) :
        _width(width), _height(height), _depth(depth),
        _cluster_buffer(cluster_buffer)
    {
        // set depth to allocate event buffer
        buffer_queue::depth = depth;

        // allocate buffer memory
        _buffer = new array_2d<buffer_queue>(width, height);
    }
    EventBuffer::~EventBuffer() {
        // deallocate buffer memory
        delete _buffer;
    }

    std::map<cid_t, ushort_t> EventBuffer::checkVicinity(const rect &domain,
        event e, ts_t tf, ts_t tc, ushort_t &num_adjacent) 
    {
        auto adjacent = std::map<cid_t, ushort_t>();

        // clip vicinity to buffer bounds
        size_t x_start = (e.x > 0) ? e.x-1U : 0;
        size_t y_start = (e.y > 0) ? e.y-1U : 0;
        // end values are not inclusive
        size_t x_end = ((e.x+2U) > _width) ? _width : e.x+2U;
        size_t y_end = ((e.y+2U) > _height) ? _height : e.y+2U;

#if USE_THREADS
        // -- lock buffers for access
        _buffer_access.lock();
#endif
        // init count to 0
        num_adjacent = 0;
        for (size_t x = x_start; x < x_end; ++x) {
            for (size_t y = y_start; y < y_end; ++y) {
                for (size_t z = 0; z < _depth; ++z) {
                    auto ts = _buffer->at(x, y)[z].t;
                    auto cid = _buffer->at(x, y)[z].cid;
                    // add to count if within filter threshold
                    if (ts > e.t-tf) num_adjacent++;
                    // add to adjacent clusters if within cluster threshold
                    if ((cid != NO_CID) && (ts > e.t-tc)) adjacent[cid]++;
                }
            }
        }
#if USE_THREADS
        // -- release buffer lock
        _buffer_access.unlock();
#endif

        // return the results
        return adjacent;
    }

    void EventBuffer::flushDomain(ts_t th, rect domain) {
        for (size_t x = domain.tl.x; x < domain.br.x; ++x) {
            for (size_t y = domain.tl.y; y < domain.br.y; ++y) {
                for (size_t z = 0; z < _depth; ++z) {
                    auto ts = _buffer->at(x, y)[z].t;
                    auto cid = _buffer->at(x, y)[z].cid;
                    // if any assigned events are too old, unassign them
                    if ((cid != NO_CID) && (ts < th)) {
#if USE_THREADS
                        // -- lock buffers for access
                        _buffer_access.lock();
#endif
                        // remove expired events from cluster buffer
                        _cluster_buffer[cid].remove(x, y);
                        // and unassign them
                        _buffer->at(x, y)[z].cid = NO_CID;
#if USE_THREADS
                        // -- release buffer lock
                        _buffer_access.unlock();
#endif
                    }
                }
            }
        }
    }

    // add event to buffer, return cid of displaced event
    void EventBuffer::addEvent(event e, cid_t cid) {
#if USE_THREADS
        // -- lock buffers for access
        _buffer_access.lock();
#endif
        // add new event to event buffer
        buffer_entry displaced = _buffer->at(e.x, e.y).push(cid, e.t);

        // add new event to cluster buffer
        if (cid != NO_CID)
            _cluster_buffer[cid].add(e.x, e.y);
        
        // remove displaced event from cluster buffer
        if (displaced.cid != NO_CID) 
            _cluster_buffer[displaced.cid].remove(e.x, e.y);
        
#if USE_THREADS
        // -- release buffer lock
        _buffer_access.unlock();
#endif
    }
};